### ISSUE TITLE
[Feature][CI] replace CI's linker with `mold`

### ISSUE THREAD

**Issue by ClSlaid** [datafuselabs/databend#4180]
**Summary**
I learned about [`mold`](https://github.com/rui314/mold) the linker from this [tweet](https://twitter.com/lichuang/status/1493156835477852163) and it turned out to be much faster and consumes much less memory than `gold`.

Here are my laptop's configurations:
```text
CPU: AMD Ryzen 7 5800H
Hard Drive: Zhitai PC005 SSD
System: Arch Linux
Kernel: Linux 5.15.22-1-lts x86_64
RAM: 13GB
rust toolchain: x86_64_unknown_linux_gnu nightly v1.60.0
```
time result:
```bash
cd databend && cargo clean

# with mold
/bin/time -v cargo test
# Command exited with non-zero status 101
#        Command being timed: "cargo test"
#        User time (seconds): 2182.67
#        System time (seconds): 160.52
#        Percent of CPU this job got: 657%
#        Elapsed (wall clock) time (h:mm:ss or m:ss): 5:56.15
#        Average shared text size (kbytes): 0
#        Average unshared data size (kbytes): 0
#        Average stack size (kbytes): 0
#        Average total size (kbytes): 0
#        Maximum resident set size (kbytes): 4888048
#        Average resident set size (kbytes): 0
#        Major (requiring I/O) page faults: 50383
#        Minor (reclaiming a frame) page faults: 41360989
#        Voluntary context switches: 352776
#        Involuntary context switches: 321932
#        Swaps: 0
#        File system inputs: 6283912
#        File system outputs: 21923208
#        Socket messages sent: 0
#        Socket messages received: 0
#        Signals delivered: 0
#        Page size (bytes): 4096
#        Exit status: 101

# with gold
cargo clean && /bin/time -v cargo test
# eating up too much swap memory, drawing down it's time benchmark

# Command exited with non-zero status 101
#        Command being timed: "cargo test"
#        User time (seconds): 2723.20
#        System time (seconds): 432.94
#        Percent of CPU this job got: 432%
#        Elapsed (wall clock) time (h:mm:ss or m:ss): 12:09.36
#        Average shared text size (kbytes): 0
#        Average unshared data size (kbytes): 0
#        Average stack size (kbytes): 0
#        Average total size (kbytes): 0
#        Maximum resident set size (kbytes): 4213836
#        Average resident set size (kbytes): 0
#        Major (requiring I/O) page faults: 5896400
#        Minor (reclaiming a frame) page faults: 74752048
#        Voluntary context switches: 6994245
#        Involuntary context switches: 306750
#        Swaps: 0
#        File system inputs: 214730184
#        File system outputs: 59830592
#        Socket messages sent: 0
#        Socket messages received: 0
#        Signals delivered: 0
#        Page size (bytes): 4096
#        Exit status: 101
```

Almost 60% reduction of time usage and 90% reduction of maximum memory consumption.

**Comment by sundy-li** [comment]
LGTM, But mold did not support MacOsX, it can be specified to just link on Linux system.

**Comment by ZhiHanZ** [comment]
zig linker also sounds good for musl/cross compile purpose，

https://github.com/messense/cargo-zigbuild

**Comment by PsiACE** [comment]
> Almost 60% reduction of time usage and 90% reduction of maximum memory consumption.

Yes, mold is good enough for linux.

> zig linker also sounds good for musl/cross compile purpose，

It seems that cargo-zigbuild is not able to complete the build of Databend locally for me and needs further investigation.

**Comment by Xuanwo** [comment]
Well, I really don't like `musl` for its poor alloc performance.

- [Why does musl make my Rust code so slow?](https://andygrove.io/2020/05/why-musl-extremely-slow/)
- [Optimizing Rust Binaries: Observation of Musl versus Glibc and Jemalloc versus System Alloc](https://users.rust-lang.org/t/optimizing-rust-binaries-observation-of-musl-versus-glibc-and-jemalloc-versus-system-alloc/8499)

**Comment by Xuanwo** [comment]
I do think `mold` is good enough to be used in our CI (as I used it locally too). @ClSlaid Would you like to try to do the replacement?

In order to make `mold` work, maybe we need to:

- Only run `mold` on Linux platform
  - maybe via env instead of using `if` in workflows (`CARGO_TARGET_<triple>_LINKER`, `CARGO_TARGET_<triple>_RUSTFLAGS`)
- Figure how to get `llvm` setup
  - maybe it is already set up or `apt install`, we need to have a try.
- Compare the difference on the same commit to make sure it really works.

**Comment by ClSlaid** [comment]
> 

I'll take a look.

**Comment by ClSlaid** [comment]
https://github.com/datafuselabs/databend/runs/5245603718?check_suite_focus=true#step:3:258
Oops, seems like the `mold` package is only available for `apt` on `ubuntu-22.04` or later versions, while GitHub only offers at most `ubuntu-20.04`.
Maybe we can try to use [`setup-mold`](https://github.com/Warchant/setup-mold) project.
